### PR TITLE
[SYCL] Enable FPGA max_concurrency memory attribute

### PR DESCRIPTION
**Pull request by vmaksimo**
Signed-off-by: Viktoria Maksimova <viktoria.maksimova@intel.com>

### CODE REVIEW

File: clang/include/clang/Basic/Attr.td [hunk 1]
```diff
@@ -1474,6 +1482,22 @@ def IntelFPGANumBanks : Attr {
   }];
 }
 
+def IntelFPGAMaxConcurrency : InheritableAttr {
+  let Spellings = [GNU<"max_concurrency">, CXX11<"intelfpga","max_concurrency">];
```
Comment (keryell) [inline clang/include/clang/Basic/Attr.td]: Perhaps `intelfpga` would be clearer if written `intel_fpga`.
Are these attributes defined to be compatible with some other tools or are just brand new attributes?

File: clang/include/clang/Basic/AttrDocs.td [hunk 2]
```diff
@@ -1709,6 +1709,17 @@ with N banks.
   }];
 }
 
+def IntelFPGAMaxConcurrencyAttrDocs : Documentation {
+  let Category = DocCatVariable;
+  let Heading = "max_concurrency (IntelFGPA)";
+  let Content = [{
+This attribute may be attached to a variable or struct member declaration and
+instructs the backend to replicate the memory generated for the variable or
```
Comment (keryell) [inline clang/include/clang/Basic/AttrDocs.td]: `memory generated` might be unclear for people not familiar with FPGA.
But I do not now what to have instead. `memory synthesized`?

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 3]
```diff
@@ -53,7 +53,22 @@ void foo1()
   //CHECK: IntelFPGANumBanksAttr
   //CHECK-NEXT: ConstantExpr
   //CHECK-NEXT: IntegerLiteral{{.*}}4{{$}}
-  __attribute__((__numbanks__(4))) unsigned int v_six2[32];
+  [[intelfpga::numbanks(4)]] unsigned int v_six2[32];
```
Comment (keryell) [inline clang/test/SemaSYCL/intel-fpga-local.cpp]: Clearer, indeed!

File: clang/test/SemaSYCL/intel-fpga-local.cpp [hunk 4]
```diff
@@ -53,7 +53,22 @@ void foo1()
...
+
+  //CHECK: VarDecl{{.*}}v_seven
+  //CHECK: IntelFPGAMemoryAttr{{.*}}Implicit
+  //CHECK: IntelFPGAMaxConcurrencyAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: IntegerLiteral{{.*}}4{{$}}
+  __attribute__((max_concurrency(4)))
```
Comment (keryell) [inline clang/test/SemaSYCL/intel-fpga-local.cpp]: I am a little bit scared about naming conflicts with thousands of top-level attributes landing from many vendors...

File: clang/include/clang/Basic/Attr.td (hunk 1 above)
Comment (MrSidims) [reply clang/include/clang/Basic/Attr.td]: > Are these attributes defined to be compatible with some other tools or are just brand new attributes?

You can already use them, for example, in Intel® FPGA SDK for OpenCL™ 19.1.

File: clang/include/clang/Basic/AttrDocs.td (hunk 2 above)
Comment (MrSidims) [reply clang/include/clang/Basic/AttrDocs.td]: > `memory generated` might be unclear for people not familiar with FPGA.
> But I do not now what to have instead. `memory synthesized`?

I agree.

File: clang/include/clang/Basic/Attr.td (hunk 1 above)
Comment (keryell) [reply clang/include/clang/Basic/Attr.td]: Interesting. With the "intelfpga::" namespace?
But honestly I have no problem about having Intel extension harder to use or unclear. ;-)

**Comment by MrSidims** [comment]
@bader @vladimirlaz why it's on 'Review required' state yet if I've already approved?

**Comment by bader** [comment]
@MrSidims I think to change the state, someone with write access has to approve aPR.

**Comment by MrSidims** [comment]
> @MrSidims I think to change the state, someone with write access has to approve a PR.

Oh, I see, I thought the special rights are about merging PR not about reviewing.

**Comment by asavonic** [comment]
@romanovvlad can you check this patch and merge if it's ok?

File: clang/test/SemaSYCL/intel-fpga-local.cpp (hunk 4 above)
Comment (asavonic) [reply clang/test/SemaSYCL/intel-fpga-local.cpp]: This is a good point. We'll think on how to solve this for C-style attributes.

**Pull request merged**